File: SimpleLinearRegression/simple_linear_regression_example.py
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
import numpy as np

# Read the CSV data
DF = pd.read_csv("example_data_simple_linear_regression.csv")

# Fit a linear model
LM = sm.OLS(DF['height'], sm.add_constant(DF['weight'])).fit()

# Create an empty canvas
fig, ax = plt.subplots(constrained_layout = True)
ax.set_xlim(0, 160)
ax.set_ylim(125, 215)
ax.set_xlabel("Weight (kg)")
ax.set_ylabel("Height (cm)")

# Add prediction band
x_values = np.arange(0, 160, 0.1)
newdata = sm.add_constant(pd.DataFrame({'weight': x_values}))
y_pred = LM.get_prediction(newdata).summary_frame()
ax.fill_between(x_values, y_pred['obs_ci_lower'], y_pred['obs_ci_upper'], 
                color = "coral", alpha = 0.5)

# Add confidence band
ax.fill_between(x_values, y_pred['mean_ci_lower'], y_pred['mean_ci_upper'], 
                color = "white")
ax.fill_between(x_values, y_pred['mean_ci_lower'], y_pred['mean_ci_upper'], 
                color = "steelblue", alpha = 0.5)

# Add the regression line
ax.plot(x_values, y_pred['mean'], color = "steelblue", linewidth = 2)

# Add the data points on top
ax.scatter(DF['weight'], DF['height'], marker = 'o', color = 'black')

# Add axes
ax.xaxis.set_ticks_position('bottom')
ax.yaxis.set_ticks_position('left')

plt.show()

File: SimpleLinearRegression/Part 1 - Example with confidence and prediction interval/simple_linear_regression_example.R
# Store this script as well as the data set in the same location, then go to:
# Session > Set Working Directory > To Source File Location
DF <- read.csv("example_data_simple_linear_regression.csv")

# Fit a linear model
LM <- lm(height ~ weight, data = DF)

# Create an empty canvas
plot(NA, bty = "n", axes = FALSE, 
     xlab = "Weight (kg)",
     ylab = "Height (cm)",
     xlim = c(0, 160), xaxs = "i",
     ylim = c(125, 215), yaxs = "i")

# Create transparent colors
coralalpha <- rgb(t(col2rgb("coral")), alpha = 128, maxColorValue = 255)
steelbluealpha <- rgb(t(col2rgb("steelblue")), alpha = 128, maxColorValue = 255)

# Add prediction band
x_values <- seq(0, 160, 0.1)
newdata  <- data.frame(weight = x_values)
y_pred   <- predict(LM, newdata, interval = "pred")
polygon(c(x_values, rev(x_values)),
        c(y_pred[, 2], rev(y_pred[, 3])), col = coralalpha, border = NA)

# Add confidence band
y_conf   <- predict(LM, newdata, interval = "conf")
polygon(c(x_values, rev(x_values)),
        c(y_conf[, 2], rev(y_conf[, 3])), col = "white", border = NA)
polygon(c(x_values, rev(x_values)),
        c(y_conf[, 2], rev(y_conf[, 3])), col = steelbluealpha, border = NA)

# Add the regression line
abline(coef(LM), lwd = 2, col = "steelblue")

# Add the data points on top
points(height ~ weight, DF, pch = 19)

# Add axes
axis(1)
axis(2, las = 1)
